docs/architecture-overview.rst

```diff
@@ -0,0 +1,297 @@
+most important of these files. It explains what happens when requests
+come in from outside.
+
+- In production, all requests to URLs beginning with ``/static/`` are
+  served from the corresponding files in ``/home/zulip/prod-static/``,
+  and the production build process (``link``) compiles, minifies,
+  installs the static assets into the ``prod-static/`` tree form. In
+  development, files are served directly from ``/static/`` in the git
+  repository.
+
+- Requests to ``/json/get_events``, ``/api/v1/events``, and
+  ``/sockjs`` are sent to the Tornado server. These are requests to
+  the real-time push system, because the user's web browser sets up a
+  long-lived TCP connection with Tornado to serve as `a channel for
+  push notifications <
+  https://en.wikipedia.org/wiki/Push_technology#Long_Polling>`__. nginx
+  gets the hostname for the Tornado server via
+  ``puppet/zulip/files/nginx/zulip-include-frontend/upstreams``.
+
+- Requests to ``/api`` and to ``/`` (i.e., everything else) are sent
+  to the Django app via the UNIX socket
+  ``unix:/home/zulip/deployments/fastcgi-socket`` (defined in
+  ``puppet/zulip/files/nginx/zulip-include-frontend/upstreams``). We
+  use ``zproject/wsgi.py`` to implement FastCGI here (see
+  ``django.core.wsgi``).
+
+
+
+Supervisor
+----------
+
+We use `supervisord <http://supervisord.org/>`__ to start server
+processes, restart them automatically if they crash, and direct
+logging.
+
+The config file is
+``zulip/puppet/zulip/files/supervisor/conf.d/zulip.conf``. This is
+where Tornado and Django are set up, as well as a number of background
+processes that process event queues. We use event queues for the kinds
+of tasks that are best run in the background because they are
+expensive (in terms of performance) and don't have to be synchronous
+-- e.g., sending emails or updating analytics. Also see :doc:`queuing`.
+
+
+memcached
+---------
+
+memcached is used to cache database model
+objects. ``zerver/lib/cache.py`` and ``zerver/lib/cache_helpers.py``
+manage putting things into memcached, and invalidating the cache when
+values change. The memcached configuration is in in
+``puppet/zulip/files/memcached.conf``.
+
+Redis
+-----
+
+Redis is used for a few very short-term data stores, such as in the
+basis of ``zerver/lib/rate_limiter.py``, a per-user rate limiting
+scheme `example
+<http://blog.domaintools.com/2013/04/rate-limiting-with-redis/>`__),
+and the `email-to-Zulip integration
+<https://zulip.com/integrations/#email>`__.
+
+Redis is configured in ``zulip/puppet/zulip/files/redis`` and it's a
+pretty standard configuration except for the last line, which turns
+off persistence:
+
+::
+
+     # Zulip-specific configuration: disable saving to disk.
+     save ""
+
+memcached was used first and then we added Redis specifically to
+implement rate limiting. `We're working on switching everything over
+to Redis.<https://github.com/zulip/zulip/issues/16>`__
```

Comment:
Would say "There's discussion of moving everything over to Redis", since I think it's an open question as to whether that actually will make sense.